### ISSUE TITLE
Review all exporters for unhandled exceptions

### ISSUE THREAD

**Issue by cijothomas** [open-telemetry/opentelemetry-dotnet#2872]
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles

"The SDK MUST NOT throw unhandled exceptions for errors in their own operations. For example, an exporter should not throw an exception when it cannot reach the endpoint to which it sends telemetry data."

This was exposed in https://github.com/open-telemetry/opentelemetry-dotnet/issues/2870 where OTLP exporter crashed the whole process, but opening a dedicated issue to track fixing this. This is separate from the [issue](https://github.com/open-telemetry/opentelemetry-dotnet/pull/1627) about when/how should SDK handle unhandled exceptions when user components behave bad. 

This issue is only tracking the core components (as required by the spec) : 

- [x] OTLPExporter - Fixed in https://github.com/open-telemetry/opentelemetry-dotnet/pull/2915
- [x] Jaeger - Fix not required. See https://github.com/open-telemetry/opentelemetry-dotnet/issues/2872#issuecomment-1036616930
- [x] Zipkin - Fixed https://github.com/open-telemetry/opentelemetry-dotnet/pull/2935
- [x] Prometheus - Fix not required. See https://github.com/open-telemetry/opentelemetry-dotnet/issues/2872#issuecomment-1043323512

**Comment by cijothomas** [comment]
@Yun-Ting for OTLPExporter. @alanwest can guide here.

**Comment by TimothyMothra** [comment]
I can help :)

**Comment by TimothyMothra** [comment]
I reviewed the Prometheus exporter.
TLDR; If this is used as designed, there should be no issues.
However, if the exporter is ever invoked while OnExport is set to null, this would throw an exception.

If we want to proceed with caution, we can add a try/catch to the Exporter or add a null check.
I'll follow up with @cijothomas offline to discuss.

---

1. `PrometheusExporter.Export` will call `PrometheusExporter.OnExport` without a null check. 
    https://github.com/open-telemetry/opentelemetry-dotnet/blob/1b5765bf43058c6301e0ff0c099aedb34541d54e/src/OpenTelemetry.Exporter.Prometheus/PrometheusExporter.cs#L76-L79

2. `PrometheusExporter.OnExport` should be set to `PrometheusCollectionManager.OnCollect` which DOES has a try/catch. 
    This eventually calls `BaseExportingMetricReader.ProcessMetrics` which will call `exporter.Export`

    https://github.com/open-telemetry/opentelemetry-dotnet/blob/1b5765bf43058c6301e0ff0c099aedb34541d54e/src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs#L177-L182

    https://github.com/open-telemetry/opentelemetry-dotnet/blob/1b5765bf43058c6301e0ff0c099aedb34541d54e/src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs#L81-L86

3. The `PrometheusCollectionManager` has a pattern I'm unfamiliar with. 
    This sets `PrometheusExporter.OnExport`, then calls `PrometheusExporter.Collect`, and then sets `PrometheusExporter.OnExport` to null.
    https://github.com/open-telemetry/opentelemetry-dotnet/blob/1b5765bf43058c6301e0ff0c099aedb34541d54e/src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs#L168-L175

**Comment by TimothyMothra** [comment]
I took a look the `JaegerExporter`. 
It's export method is covered with a try/catch.
https://github.com/open-telemetry/opentelemetry-dotnet/blob/6ca4c13a188585c00c8655e2749e30f5d96cf41a/src/OpenTelemetry.Exporter.Jaeger/JaegerExporter.cs#L95-L98

@cijothomas, I think this is safe and can be marked as complete above :)

**Comment by cijothomas** [comment]
https://github.com/open-telemetry/opentelemetry-dotnet/issues/2905 another crash from OTLP.
